lib/registration.cpp

```diff
@@ -13,4 +20,309 @@ transform_pointcloud_inplace(EigenPointCloudRef cloud,
+{
+  assert(i >= 0 && i < size_);
+
+  while (i != subsets_[i]) {
+    subsets_[i] = subsets_[subsets_[i]];
+    i = subsets_[i];
+  }
+
+  return i;
+}
+
+IndexType
+DisjointSet::Union(IndexType i, IndexType j, bool balance_sizes)
+{
+  // Ensure i and j are valid indices
+  assert(i >= 0 && i < size_);
+  assert(j >= 0 && j < size_);
+
+  // Find the root representatives of the subsets containing i and j
+  IndexType root_i = Find(i);
+  IndexType root_j = Find(j);
+
+  if (root_i != root_j) {
+    if (balance_sizes) {
+      // If balance_sizes is true, merge the larger subset into the smaller one
+      if (numbers_[root_i] > numbers_[root_j]) {
+        numbers_[root_j] += numbers_[root_i];
+        subsets_[root_i] = root_j;
+        return root_j;
+      } else {
+        numbers_[root_i] += numbers_[root_j];
+        subsets_[root_j] = root_i;
+
+        return root_i;
+      }
+    } else {
+      // Always merge i's subset into j's subset
+      numbers_[root_j] += numbers_[root_i];
+      subsets_[root_i] = root_j;
+      numbers_[root_i] = 0;
+      return root_j;
+    }
+  }
+
+  // If the subsets already have the same root representative, no action is
+  // needed
+  return root_i;
+}
+
+std::size_t
+estimate_supervoxel_count(EigenPointCloudConstRef cloud, double seed_resolution)
+{
+  // Calculate the point cloud point bounding box
+  using Coordinate = Eigen::Array<double, 1, 3>;
+  Coordinate lowerleft = cloud.colwise().minCoeff();
+  Coordinate upperright = cloud.colwise().maxCoeff();
+
+  // Subdivide the bounding box according to the given seed_resolution
+  using std::floor;
+  using MultiIndex = Eigen::Array<std::size_t, 1, 3>;
+  MultiIndex voxelsizes =
+    floor(((upperright - lowerleft) / seed_resolution) + 1)
+      .cast<std::size_t>()
+      .eval();
+
+  // Assign all points to their voxels by throwing them in a hashmap
+  std::unordered_map<std::size_t, std::vector<Eigen::Index>> voxelmap;
+  for (IndexType i = 0; i < cloud.rows(); ++i) {
+    Coordinate coord(cloud.row(i));
+    auto ind = floor(((coord - lowerleft) / seed_resolution) + 1)
+                 .cast<std::size_t>()
+                 .eval();
+    voxelmap[ind[2] * voxelsizes[1] * voxelsizes[0] + ind[1] * voxelsizes[0] +
+             ind[0]]
+      .push_back(i);
+  }
+
+  return voxelmap.size();
+}
+
+double
+squaredEuclideanDistance(const Eigen::RowVector3d& point1,
+                         const Eigen::RowVector3d& point2)
+{
+  const auto diff = point1 - point2;
+  return diff.squaredNorm();
+}
+
+std::vector<std::vector<int>>
+supervoxel_segmentation(Epoch& epoch,
+                        const KDTree& kdtree, //?
+                        double resolution,
+                        int k)
+{
+
+  // Define number of supervoxels and labels.
+  auto n_supervoxels = estimate_supervoxel_count(epoch.cloud, resolution);
+  std::vector<int> labels(epoch.cloud.rows(), -1);
+  DisjointSet set(epoch.cloud.rows());
+
+  // Calculate normals for vccs metric
```

Comment:
With normal calculation being available on the Python API, I think it would possible to omit this calculation here and instead trigger it from the Python side if and only if they are not already present. As is, I think we would recompute existing normals (which is ridiculously expensive).